api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier.
+// This is a dummy class that is meant to be overridden,
+// the methods return default values.
+class Context
+{
+
+public:
+  // The Key class is used to obscure access from the
+  //  user to the context map. The identifier is used as a key
+  // to the context map.
+  class Key
+  {
+
+  public:
+    // Returns the key's identifier.
+    nostd::string_view GetIdentifier() { return ""; }
+
+    // Returns the key's name
+    nostd::string_view GetName() { return ""; }
+
+    Key() = default;
+
+    // Constructs a new Key with the passed in name. Sets the identifier as
+    // the address of this object.
+    Key(nostd::string_view key_name) {}
+
+  private:
+    friend class Context;
+  };
+
+  // Creates a key with the passed in name and returns it.
+  Key CreateKey(nostd::string_view key_name) { return Key(key_name); }
+
+  Context() = default;
+
+  // Contructor, creates a context object from a map of keys
+  // and identifiers.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &attributes)
+  {}
+
+  // Accepts a new iterable and then returns a new  context that
+  // contains both the original pairs and the new pair.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context WriteValues(T &attributes) noexcept
+  {
+    return Context(attributes);
+  }
+
+  // Returns the value associated with the passed in key.
+  virtual common::AttributeValue GetValue(Key key) { return 0; }
```

Comment:
It would be more elegant and in C++ spirit to use `[]` operator and `nostd::string_view` to obtain the key, as it's done in a map. Since we're dealing with dictionary, it'd be more practical this dictionary to have `std::map`-alike properties, like access operators.

Reply:
Map operator[] returns a T& (non-const) and constructs an element dynamically if it's not already present. 

@maxgolov Are those behavior you want for Context?

Reply:
@0x4b - but this empty default also returns 0 for non-existing keys, and doesn't describe expected behavior in case if attribute does not exist in CorrelationContext, for example. We could say that if the key does not exist - we either throw (std::out_of_range-alike exception) or return a NULL object. We can allow a special AttributeValue of null, in that case the operator can return a reference to constructed NULL object.

Reply:
I didn't mean my question to be rhetorical. I don't know what behavior OT wants for value lookup? Returning the 

[In another PR you asked for support for environments that don't support exceptions](https://github.com/open-telemetry/opentelemetry-cpp/pull/143#discussion_r450498669), so it seems like there needs to be an option other than throwing here. The options I see are

- optional<>, but I'm assuming you're not interested in a nostd::optional.  Is that right? 
- a [valueless](https://en.cppreference.com/w/cpp/utility/variant/valueless_by_exception) [no]std::variant (assuming that's the value type). that feels kind of unnatural
- a default-constructed AttributeValue (i.e. false bool variant)
- a pointer to the value type

Whether to use string_view for the key is it's own question, and there's also an open question about what the value type should be, so maybe that discussion will resolve in a way that implies what an unknown value should be.

Reply:
I think we can discuss:
- valueless `nostd::variant` as NULL or Empty object as a preference for both exceptions and no-exceptions build -  provides consistent behavior for both flavors.
- throw on exceptions build, and THROW macro as a macro == `std::terminate` on no-exceptions build - less preferred, since API behavior would be different.

I shared some code below providing an implementation with both suggestions, including a placeholder for NULL-object in a variant. Default constructed Value should be NULL-or-empty object, I think it's gonna be fairly natural. It's empty until you assign something to it.